### ISSUE TITLE
Race condition between Snapshot deletion and commit

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3764]
**Describe the bug**

It seems that there exist an race condition between deleting a pending snapshot and commiting one.

```
I 2020-01-30T02:38:52.000704565Z 2020-01-30 02:38:52.000 [] [Broker-2-StateReplication-1] INFO  io.zeebe.broker.clustering.atomix.storage.snapshot.AtomixSnapshotStorage - Max snapshot count reached (4), purging snapshots older than DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/39548516-3-1580349227165-98002571440712, metadata=DbSnapshotMetadata{index=39548516, term=3, timestamp=2020-01-30 01:53:47,165, position=98002571440712}}
 
I 2020-01-30T02:38:52.000893185Z 2020-01-30 02:38:52.000 [] [Broker-2-StateReplication-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Deleting snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/38549092-3-1580348330845-95584499626088, metadata=DbSnapshotMetadata{index=38549092, term=3, timestamp=2020-01-30 01:38:50,845, position=95584499626088}}
 
I 2020-01-30T02:38:52.039031563Z 2020-01-30 02:38:52.038 [] [Broker-2-StateReplication-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.AtomixSnapshotStorage - Snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/38549092-3-1580348330845-95584499626088, metadata=DbSnapshotMetadata{index=38549092, term=3, timestamp=2020-01-30 01:38:50,845, position=95584499626088}} removed from store io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore@1c1c1850
 
I 2020-01-30T02:38:52.039060737Z 2020-01-30 02:38:52.038 [] [Broker-2-StateReplication-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Search for orphaned snapshots below oldest valid snapshot DbSnapshotMetadata{index=39548516, term=3, timestamp=2020-01-30 01:53:47,165, position=98002571440712}
 
I 2020-01-30T02:38:52.111263446Z 2020-01-30 02:38:52.111 [] [Broker-2-StateReplication-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Delete not completed (orphaned) snapshot /usr/local/zeebe/data/raft-partition/partitions/1/pending/42497265-5-1580351919842
 
I 2020-01-30T02:38:52.111891267Z 2020-01-30 02:38:52.111 [] [Broker-2-StateReplication-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Committed new snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/42497265-5-1580351919842-105136510471232, metadata=DbSnapshotMetadata{index=42497265, term=5, timestamp=2020-01-30 02:38:39,842, position=105136510471232}}
 
I 2020-01-30T02:38:52.540714258Z 2020-01-30 02:38:52.540 [Broker-2-DeletionService-1] [Broker-2-zb-actors-1] DEBUG io.zeebe.broker.system - The last processed position for partition 1 is 98028335531928
 
I 2020-01-30T02:38:52.541701198Z 2020-01-30 02:38:52.541 [Broker-2-DeletionService-1] [Broker-2-zb-actors-1] DEBUG io.zeebe.broker.system - The lowest exported position for partition 1 is 98024045558592
 
I 2020-01-30T02:38:53.386594763Z 2020-01-30 02:38:53.385 [] [raft-server-2-raft-partition-partition-1] WARN  io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot - Failed to delete pending snapshot DbPendingSnapshot{index=42497265, term=5, timestamp=2020-01-30 02:38:39,842, directory=/usr/local/zeebe/data/raft-partition/partitions/1/pending/42497265-5-1580351919842}
 
I 2020-01-30T02:38:53.386622010Z java.nio.file.NoSuchFileException: /usr/local/zeebe/data/raft-partition/partitions/1/pending/42497265-5-1580351919842
	at sun.nio.fs.UnixException.translateToIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.readAttributes(Unknown Source) ~[?:?]
	at sun.nio.fs.LinuxFileSystemProvider.readAttributes(Unknown Source) ~[?:?]
	at java.nio.file.Files.readAttributes(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.getAttributes(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.visit(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.walk(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at io.zeebe.util.FileUtil.deleteFolder(FileUtil.java:41) ~[zeebe-util-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot.abort(DbPendingSnapshot.java:120) ~[zeebe-broker-0.23.0-SNAPSHOT.jar:0.23.0-SNAPSHOT]
	at io.atomix.protocols.raft.roles.PassiveRole.abortPendingSnapshot(PassiveRole.java:521) ~[atomix-raft-3.2.0-20200128.142312-118.jar:?]
	at io.atomix.protocols.raft.roles.PassiveRole.onInstall(PassiveRole.java:249) ~[atomix-raft-3.2.0-20200128.142312-118.jar:?]
	at io.atomix.protocols.raft.roles.FollowerRole.onInstall(FollowerRole.java:81) ~[atomix-raft-3.2.0-20200128.142312-118.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.lambda$null$10(RaftContext.java:218) ~[atomix-raft-3.2.0-20200128.142312-118.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.lambda$runOnContext$33(RaftContext.java:255) ~[atomix-raft-3.2.0-20200128.142312-118.jar:?]
	at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-20200128.142229-118.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
 
```

We can see in the logs above that the snapshot `42497265-5-1580351919842` is pending and need to be deleted but directly after that committed, which means the deletion fails.

```

I 2020-01-30T02:38:52.111263446Z 2020-01-30 02:38:52.111 [] [Broker-2-StateReplication-1] DEBUG io.zeebe.broker.clustering.atomix.storage.snapshot.DbSnapshotStore - Delete not completed (orphaned) snapshot /usr/local/zeebe/data/raft-partition/partitions/1/pending/42497265-5-1580351919842
Committed new snapshot DbSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/42497265-5-1580351919842-105136510471232, metadata=DbSnapshotMetadata{index=42497265, term=5, timestamp=2020-01-30 02:38:39,842, position=105136510471232}}
...

I 2020-01-30T02:38:53.386594763Z 2020-01-30 02:38:53.385 [] [raft-server-2-raft-partition-partition-1] WARN  io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot - Failed to delete pending snapshot DbPendingSnapshot{index=42497265, term=5, timestamp=2020-01-30 02:38:39,842, directory=/usr/local/zeebe/data/raft-partition/partitions/1/pending/42497265-5-1580351919842}
 
```

**Comment by npepinpe** [comment]
Correct me if I'm wrong, but here we're trying to delete pending snapshots older than `38549092-3-1580348330845-95584499626088` (taken at 01/30/2020 @ 1:38am (UTC)) - and we end up deleting `42497265-5-1580351919842-105136510471232` (taken at 01/30/2020 @ 2:38am (UTC)), which is clearly newer? :scream:

**Comment by npepinpe** [comment]
So, my bad, seems the condition was inverted and we were deleting pending snapshots newer than the cutoff :see_no_evil: So sorry about that

**Comment by npepinpe** [comment]
That said, I can see how we might still have the race condition (though it should become less likely). As committing and deleting is done in separate threads, it's possible for this to happen anyway. Considering that (with the fix :see_no_evil:) we delete snapshots which cannot be committed anyway, deleting should win over committing - so in that sense we shouldn't let the orphaned snapshot be committed (if possible).

**Comment by npepinpe** [comment]
Like I said, this isn't fixed per se - it's possible to commit a pending snapshot while we're deleting it - but this was made likely due to the bug that was fixed. With the fix this should be unlikely depending on your snapshot interval. So I'll leave it open, but it's possible we stop seeing those logs.

**Comment by npepinpe** [comment]
Should be fixed by #4113
